libcaf_net/test/scribe.cpp

```diff
@@ -0,0 +1,182 @@
+                                                            &manager.system(),
+                                                            cfg, &manager);
+    anon_send(listener, resolve_atom::value, std::move(path), p);
+  }
+
+  template <class Transport>
+  void timeout(Transport&, atom_value, uint64_t) {
+    // nop
+  }
+
+  void handle_error(sec) {
+    // nop
+  }
+
+  static expected<std::vector<char>> serialize(actor_system& sys,
+                                               const type_erased_tuple& x) {
+    std::vector<char> result;
+    binary_serializer sink{sys, result};
+    if (auto err = message::save(sink, x))
+      return err;
+    return result;
+  }
+
+private:
+  std::shared_ptr<std::vector<char>> rec_buf_;
+};
+
+} // namespace
+
+CAF_TEST_FIXTURE_SCOPE(endpoint_manager_tests, fixture)
+
+CAF_TEST(receive) {
+  std::vector<char> read_buf(1024);
+  CAF_CHECK_EQUAL(mpx->num_socket_managers(), 1u);
+  auto buf = std::make_shared<std::vector<char>>();
+  auto sockets = unbox(make_stream_socket_pair());
+  nonblocking(sockets.second, true);
+  CAF_CHECK_EQUAL(read(sockets.second, read_buf.data(), read_buf.size()),
+                  sec::unavailable_or_would_block);
+  auto guard = detail::make_scope_guard([&] { close(sockets.second); });
+
+  CAF_MESSAGE("configure scribe_policy");
+  policy::scribe scribe{sockets.first};
+  scribe.configure_read(net::receive_policy::exactly(hello_manager.size()));
+  auto mgr = make_endpoint_manager(mpx, sys, scribe, dummy_application{buf});
+  CAF_CHECK_EQUAL(mgr->init(), none);
+  mpx->handle_updates();
+  CAF_CHECK_EQUAL(mpx->num_socket_managers(), 2u);
+
+  CAF_MESSAGE("sending data to scribe_policy");
+  CAF_CHECK_EQUAL(write(sockets.second, hello_manager.data(),
+                        hello_manager.size()),
+                  hello_manager.size());
+  run();
+  CAF_CHECK_EQUAL(string_view(buf->data(), buf->size()), hello_manager);
+}
+
+CAF_TEST(resolve and proxy communication) {
+  std::vector<char> read_buf(1024);
+  auto buf = std::make_shared<std::vector<char>>();
+  auto sockets = unbox(make_stream_socket_pair());
+  nonblocking(sockets.second, true);
+  auto guard = detail::make_scope_guard([&] { close(sockets.second); });
+  auto mgr = make_endpoint_manager(mpx, sys, policy::scribe{sockets.first},
+                                   dummy_application{buf});
+  CAF_CHECK_EQUAL(mgr->init(), none);
+  mpx->handle_updates();
+  run();
+  mgr->resolve("/id/42", self);
+  run();
+  self->receive(
+    [&](resolve_atom, const std::string&, const strong_actor_ptr& p) {
+      CAF_MESSAGE("got a proxy, send a message to it");
+      self->send(actor_cast<actor>(p), "hello proxy!");
+    },
+    after(std::chrono::seconds(0)) >>
+      [&] { CAF_FAIL("manager did not respond with a proxy."); });
+  run();
+  auto read_res = read(sockets.second, read_buf.data(), read_buf.size());
+  if (!holds_alternative<size_t>(read_res)) {
```

Comment:
Instead of `CAF_ERROR` and returning, we can simply use `CAF_FAIL` instead:

```c++
  if (!holds_alternative<size_t>(read_res))
    CAF_FAIL("read() returned an error: " << sys.render(get<sec>(read_res)));
```